Introduce sync_cross_rank_decision

Microve opened the pull request:
Summary:
This is an improvement over `_broadcast_rank0_decision` where we uses the rank0's decision to broadcast to every rank. The issue of `_broadcast_rank0_decision` is that we observed large variance on the peak memory usage. One cause is that different ranks receive different dynamic shaped tensors and the hints of those tensors are different in different ranks. If we only rely on rank0's decision and it's unlucky to get unrepresentative hints, then the decision it makes may not be suitable for other ranks.

Here, we introduce `sync_cross_rank_decision` which comes up with the decision after comparing all ranks' local decision, it will:
1. all gather decisions from all ranks;
2. test each decision on the current rank and get its estimated memory usage;
3. all reduce estimated memory usage with ReduceOp.MAX, so that we know the maximum memory usage of each decision on all ranks;
4. pick the decision which gives us minimum maximum memory memory usage;

A graph to show more details
https://internalfb.com/excalidraw/EX484509

After applying sync_cross_rank_decision, we observed that the variance are much smaller

Rollback Plan:

Differential Revision: D76714005


cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k

Microve commented on torch/_functorch/partitioners.py:
```diff
@@ -2476,11 +2476,27 @@ def has_same_nodes(joint_graph):
+            torch.distributed.all_reduce(saved_sizes_tensor, op=torch.distributed.distributed_c10d.ReduceOp.MAX)
+
+            picked_rank_idx = torch.argmin(saved_sizes_tensor).item()
+            log.info(f"{picked_rank_idx=}, {saved_sizes_tensor=}")
```
Sure, updated to `log.debug`

xuanzhang816 commented:
@pytorchbot label "topic: not user facing"

fmassa reviewed (approved):
Thanks!

bdhirsh commented on torch/_functorch/partitioners.py:
```diff
@@ -2515,11 +2515,38 @@ def has_same_nodes(joint_graph):
+                saved_sizes.append(saved_size)
+
+            saved_sizes_tensor = torch.tensor(
+                saved_sizes, device=torch.cuda.current_device()
```
Instead of hardcoding the cuda device here, can we try to use the same device that our collectives are going to use?

Maybe @wconstab knows of a cleaner way, but it looks like `all_gather_object` figures out what device to use with [this](https://github.com/pytorch/pytorch/blob/main/torch/distributed/distributed_c10d.py#L3153C5-L3153C52)
```
current_device = _get_object_coll_device(_group_or_default_group())
```

Microve committed:
Introduce _sync_decision_cross_ranks (#156287)

Summary:

This is an improvement over `_broadcast_rank0_decision` where we uses the rank0's decision to broadcast to every rank. The issue of `_broadcast_rank0_decision` is that we observed large variance on the peak memory usage. One cause is that different ranks receive different dynamic shaped tensors and the hints of those tensors are different in different ranks. If we only rely on rank0's decision and it's unlucky to get unrepresentative hints, then the decision it makes may not be suitable for other ranks.

Here, we introduce `_sync_decision_cross_ranks` which comes up with the decision after comparing all ranks' local decision, it will:
1. all gather decisions from all ranks;
2. test each decision on the current rank and get its estimated memory usage;
3. all reduce estimated memory usage with ReduceOp.MAX, so that we know the maximum memory usage of each decision on all ranks;
4. pick the decision which gives us minimum maximum memory usage;

A graph to show more details
https://internalfb.com/excalidraw/EX484509

After applying `_sync_decision_cross_ranks`, we observed that the variance are much smaller

Test Plan:
5 runs with `_broadcast_rank0_decision` and `_sync_decision_cross_ranks`
|   |Avg of Peak GPU Memory Utilization  |  Standard Variance of Peak GPU Memory Utilization |
| -- |
| _broadcast_rank0_decision  | 78.064%  | 0.031 |
| _sync_decision_cross_ranks | 71.838% | 0.005   |

Raw results:
`_broadcast_rank0_decision`
|  | Peak GPU Memory Utilization|
| -- |
|run1| 81.25%|
|run2| 73.75%|
|run3| 79%|
|run4| 81.2%|
|run5| 75.12%|
`_sync_decision_cross_ranks`
|  | Peak GPU Memory Utilization|
| -- |
|run1| 72.31%|
|run2| 72.47%|
|run3| 71.72%|
|run4| 71.56%|
|run5| 71.13%|

Rollback Plan:

Reviewed By: xuanzhang816, fmassa

Differential Revision: D76714005

bdhirsh reviewed (approved):
sgtm
